File: app/adapters/ollama_native.py
"""Ollama's native /api/chat endpoint (NDJSON streaming, not SSE).

This is the one adapter that gets ground-truth server-side timings: the
final line of the stream carries load_duration, prompt_eval_duration, and
eval_duration in nanoseconds. Comparing those against client-measured wall
time isolates network/proxy overhead from actual model compute -- useful
when you also have the same model running behind OpenWebUI and want to know
what the proxy costs you.
"""
from __future__ import annotations

import json
from typing import AsyncIterator

import httpx

from app.adapters.base import Adapter
from app.models import Endpoint, PromptCase, StreamEvent

NS_PER_S = 1_000_000_000


def _headers(endpoint: Endpoint) -> dict[str, str]:
    headers = {"Content-Type": "application/json"}
    if endpoint.api_key:
        headers["Authorization"] = f"Bearer {endpoint.api_key}"
    return headers


def _payload(endpoint: Endpoint, case: PromptCase) -> dict:
    messages = []
    if case.system_prompt:
        messages.append({"role": "system", "content": case.system_prompt})
    messages.append({"role": "user", "content": case.user_prompt})
    options: dict = {"temperature": case.temperature}
    if case.seed is not None:
        options["seed"] = case.seed
    if case.max_tokens is not None:
        options["num_predict"] = case.max_tokens
    return {
        "model": endpoint.model,
        "messages": messages,
        "stream": True,
        "options": options,
    }


class OllamaNativeAdapter(Adapter):
    async def stream_chat(
        self,
        client: httpx.AsyncClient,
        endpoint: Endpoint,
        case: PromptCase,
    ) -> AsyncIterator[StreamEvent]:
        base = endpoint.base_url.rstrip("/")
        url = base + "/api/chat"
        headers = _headers(endpoint)
        payload = _payload(endpoint, case)

        try:
            async with client.stream("POST", url, headers=headers, json=payload) as resp:
                if resp.status_code >= 400:
                    body = await resp.aread()
                    yield StreamEvent(
                        kind="error",
                        error_message=f"HTTP {resp.status_code}: {body[:500]!r}",
                    )
                    return

                async for line in resp.aiter_lines():
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        obj = json.loads(line)
                    except json.JSONDecodeError:
                        continue

                    if obj.get("error"):
                        yield StreamEvent(kind="error", error_message=str(obj["error"]))
                        return

                    message = obj.get("message") or {}
                    text = message.get("content") or ""
                    if text:
                        yield StreamEvent(kind="delta", text=text)

                    if obj.get("done"):
                        load_ns = obj.get("load_duration")
                        prompt_ns = obj.get("prompt_eval_duration")
                        eval_ns = obj.get("eval_duration")
                        yield StreamEvent(
                            kind="done",
                            prompt_tokens=obj.get("prompt_eval_count"),
                            completion_tokens=obj.get("eval_count"),
                            server_load_s=(load_ns / NS_PER_S) if load_ns is not None else None,
                            server_prompt_eval_s=(prompt_ns / NS_PER_S) if prompt_ns is not None else None,
                            server_eval_s=(eval_ns / NS_PER_S) if eval_ns is not None else None,
                        )
                        return

                yield StreamEvent(kind="done")
        except httpx.HTTPError as exc:
            yield StreamEvent(kind="error", error_message=f"request failed: {exc}")

    async def list_models(self, client: httpx.AsyncClient, endpoint: Endpoint) -> list[str]:
        base = endpoint.base_url.rstrip("/")
        headers = _headers(endpoint)
        resp = await client.get(base + "/api/tags", headers=headers)
        resp.raise_for_status()
        data = resp.json()
        return [m.get("name", "") for m in data.get("models", []) if m.get("name")]

File: lib/adapters/ollamaNative.ts
// Ollama's native /api/chat endpoint (NDJSON streaming, not SSE).
//
// This is the one adapter that gets ground-truth server-side timings: the
// final line of the stream carries load_duration, prompt_eval_duration,
// and eval_duration in nanoseconds. Comparing those against client-measured
// wall time isolates network/proxy overhead from actual model compute --
// useful when the same model is also reachable through OpenWebUI and you
// want to know what the proxy costs you.

import type { Adapter } from "@/lib/adapters/base";
import { authHeaders, streamLines } from "@/lib/adapters/base";
import type { Endpoint, PromptCase, StreamEvent } from "@/lib/types";

const NS_PER_S = 1_000_000_000;

function buildPayload(endpoint: Endpoint, promptCase: PromptCase) {
  const messages: { role: string; content: string }[] = [];
  if (promptCase.system_prompt) messages.push({ role: "system", content: promptCase.system_prompt });
  messages.push({ role: "user", content: promptCase.user_prompt });

  const options: Record<string, unknown> = { temperature: promptCase.temperature };
  if (promptCase.seed != null) options.seed = promptCase.seed;
  if (promptCase.max_tokens != null) options.num_predict = promptCase.max_tokens;

  return { model: endpoint.model, messages, stream: true, options };
}

async function* streamChat(endpoint: Endpoint, promptCase: PromptCase): AsyncGenerator<StreamEvent> {
  const base = endpoint.base_url.replace(/\/+$/, "");
  const headers = authHeaders(endpoint);
  const payload = buildPayload(endpoint, promptCase);

  let resp: Response;
  try {
    resp = await fetch(base + "/api/chat", { method: "POST", headers, body: JSON.stringify(payload) });
  } catch (exc) {
    yield { kind: "error", error_message: `request failed: ${exc}` };
    return;
  }

  if (resp.status >= 400) {
    const body = await resp.text();
    yield { kind: "error", error_message: `HTTP ${resp.status}: ${body.slice(0, 500)}` };
    return;
  }

  for await (const line of streamLines(resp)) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let obj: any;
    try {
      obj = JSON.parse(trimmed);
    } catch {
      continue;
    }

    if (obj.error) {
      yield { kind: "error", error_message: String(obj.error) };
      return;
    }

    const text = obj.message?.content;
    if (text) yield { kind: "delta", text };

    if (obj.done) {
      const loadNs = obj.load_duration;
      const promptNs = obj.prompt_eval_duration;
      const evalNs = obj.eval_duration;
      yield {
        kind: "done",
        prompt_tokens: obj.prompt_eval_count ?? null,
        completion_tokens: obj.eval_count ?? null,
        server_load_s: loadNs != null ? loadNs / NS_PER_S : null,
        server_prompt_eval_s: promptNs != null ? promptNs / NS_PER_S : null,
        server_eval_s: evalNs != null ? evalNs / NS_PER_S : null,
      };
      return;
    }
  }

  yield {
    kind: "done",
    prompt_tokens: null,
    completion_tokens: null,
    server_load_s: null,
    server_prompt_eval_s: null,
    server_eval_s: null,
  };
}

async function listModels(endpoint: Endpoint): Promise<string[]> {
  const base = endpoint.base_url.replace(/\/+$/, "");
  const headers = authHeaders(endpoint);
  const resp = await fetch(base + "/api/tags", { headers });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  const data = await resp.json();
  return (data.models ?? []).map((m: any) => m.name).filter(Boolean);
}

export const ollamaNativeAdapter: Adapter = { streamChat, listModels };
